File: OMGFormURLEncode.m
#import <Foundation/Foundation.h>
#import "OMGFormURLEncode.h"

static inline NSString *enc(NSString *in) {
	return (__bridge_transfer  NSString *) CFURLCreateStringByAddingPercentEscapes(
            kCFAllocatorDefault,
            (__bridge CFStringRef)in.description,
            CFSTR("[]."),
            CFSTR(":/?&=;+!@#$()',*"),
            kCFStringEncodingUTF8);
}

static NSArray *DoQueryMagic(NSString *key, id value) {
    NSMutableArray *parts = [NSMutableArray new];

    // Sort dictionary keys to ensure consistent ordering in query string,
    // which is important when deserializing potentially ambiguous sequences,
    // such as an array of dictionaries
    #define sortDescriptor [NSSortDescriptor sortDescriptorWithKey:@"description" ascending:YES selector:@selector(compare:)]

    if ([value isKindOfClass:[NSDictionary class]]) {
        NSDictionary *dictionary = value;
        for (id nestedKey in [dictionary.allKeys sortedArrayUsingDescriptors:@[sortDescriptor]]) {
            id recursiveKey = key ? [NSString stringWithFormat:@"%@[%@]", key, nestedKey] : nestedKey;
            [parts addObjectsFromArray:DoQueryMagic(recursiveKey, dictionary[nestedKey])];
        }
    } else if ([value isKindOfClass:[NSArray class]]) {
        for (id nestedValue in value)
            [parts addObjectsFromArray:DoQueryMagic([NSString stringWithFormat:@"%@[]", key], nestedValue)];
    } else if ([value isKindOfClass:[NSSet class]]) {
        for (id obj in [value sortedArrayUsingDescriptors:@[sortDescriptor]])
            [parts addObjectsFromArray:DoQueryMagic(key, obj)];
    } else {
        [parts addObjectsFromArray:[NSArray arrayWithObjects:key, value, nil]];
    }

    return parts;

    #undef sortDescriptor
}



NSString *OMGFormURLEncode(NSDictionary *parameters) {
    if (parameters.count == 0)
        return nil;
    NSMutableString *queryString = [NSMutableString new];
    NSEnumerator *e = DoQueryMagic(nil, parameters).objectEnumerator;
    for (;;) {
        id obj = e.nextObject;
        if (!obj) break;
        [queryString appendFormat:@"%@=%@&", enc(obj), enc(e.nextObject)];
    }
    [queryString deleteCharactersInRange:NSMakeRange(queryString.length - 1, 1)];
    return queryString;
}

File: Sources/OMGFormURLEncode.swift
import Foundation

private func enc(input: String, _ ignore: String) -> String {
    let allowedSet = NSMutableCharacterSet(charactersInString: ignore)
    allowedSet.formUnionWithCharacterSet(.URLQueryAllowedCharacterSet())
    allowedSet.removeCharactersInString(":/?&=;+!@#$()',*")
    return input.stringByAddingPercentEncodingWithAllowedCharacters(allowedSet)!
}

private func DoQueryMagic(key: String?, value: NSObject) -> [String] {
    var parts = [String]()
    // Sort dictionary keys to ensure consistent ordering in query string,
    // which is important when deserializing potentially ambiguous sequences,
    // such as an array of dictionaries
    if let dictionary = value as? [String: NSObject] {
        for nestedKey in dictionary.keys.sort() {
            let recursiveKey = key != nil ? "\(key!)[\(nestedKey)]" : nestedKey
            parts += DoQueryMagic(recursiveKey, value: dictionary[nestedKey]!)
        }
    } else if let array = value as? [NSObject] {
        for nestedValue in array {
            parts += DoQueryMagic("\(key ?? "")[]", value: nestedValue)
        }
    } else if let set = value as? Set<NSObject> {
        for obj in set.sort( { $0.description < $1.description } ) {
            parts += DoQueryMagic(key, value: obj)
        }
    } else if let key = key {
        parts.append(key)
        parts.append(value.description)
    }
    return parts
}

/**
 Express this dictionary as a `application/x-www-form-urlencoded` string.
 
 Most users would recognize the result of this transformation as the query
 string in a browser bar. For our purposes it is the query string in a GET
 request and the HTTP body for POST, PUT and DELETE requests.
 
 If the parameters dictionary is nil or empty, returns nil.
*/
public func OMGFormURLEncode(parameters: [String: NSObject]) -> String {
    guard parameters.count > 0 else {
        return ""
    }
    var queryString = String()
    let array = DoQueryMagic(nil, value: parameters)
    var e = array.generate()
    while true {
        guard let obj = e.next() else {
            break
        }
        queryString += "\(enc(obj, "[]"))=\(enc(e.next() ?? "", ""))&"
    }
    return String(queryString.characters.dropLast())
}
